🐛 [Bug]: cache middleware: runtime error: index out of range [0] with length 0

brunodmartins opened the pull request:
# Description

Fixes #3072 

The bug happens due to some scenarios together: 

- The cache middleware always create a pointer `item` even though the cache entry was not found. 
- The application can define a `CacheInvalidator` that commands to expire the cache **even though** it doesn't not have any entry

In case the application defines a `MaxBytes` size for the cache, it will try to remove it from the heap, resulting in a runtime error due to the heap not having even initialized.

To fix the issue, I made the following changes:

- The cache will return `nil` in case it does not found the entry. 
- The expiration flow will not be triggered if no cache entry was found, this will avoid unexpected behaviors.


## Changes introduced

List the new features or adjustments introduced in this pull request. Provide details on benchmarks, documentation updates, changelog entries, and if applicable, the migration guide.

- [ ] Benchmarks: Describe any performance benchmarks and improvements related to the changes.
- [ ] Documentation Update: Detail the updates made to the documentation and links to the changed files.
- [ ] Changelog/What's New: Include a summary of the additions for the upcoming release notes.
- [ ] Migration Guide: If necessary, provide a guide or steps for users to migrate their existing code to accommodate these changes.
- [ ] API Alignment with Express: Explain how the changes align with the Express API.
- [ ] API Longevity: Discuss the steps taken to ensure that the new or updated APIs are consistent and not prone to breaking changes.
- [ ] Examples: Provide examples demonstrating the new features or changes in action.

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (improvement to existing features and functionality)
- [ ] Documentation update (changes to documentation)
- [ ] Performance improvement (non-breaking change which improves efficiency)
- [x] Code consistency (non-breaking change which improves code reliability and robustness)

## Checklist

Before you submit your pull request, please make sure you meet these requirements:

- [ ] Followed the inspiration of the Express.js framework for new functionalities, making them similar in usage.
- [x] Conducted a self-review of the code and provided comments for complex or critical parts.
- [ ] Updated the documentation in the `/docs/` directory for [Fiber's documentation](https://docs.gofiber.io/).
- [x] Added or updated unit tests to validate the effectiveness of the changes or new features.
- [x] Ensured that new and existing unit tests pass locally with the changes.
- [ ] Verified that any new dependencies are essential and have been agreed upon by the maintainers/community.
- [x] Aimed for optimal performance with minimal allocations in the new code.
- [ ] Provided benchmarks for the new code to analyze and improve upon.

## Commit formatting

Please use emojis in commit messages for an easy way to identify the purpose or intention of a commit. Check out the emoji cheatsheet here: [CONTRIBUTING.md](https://github.com/gofiber/fiber/blob/master/.github/CONTRIBUTING.md#pull-requests-or-commits)

ReneWerner87 commented:
@brunodmartins thanks for the fix, can you please create another unittests for the case from the bug ticket so that we have this case covered forever

this unittest should lead to errors in the master and go through successfully in a pull request if you would transfer it for testing purposes

brunodmartins commented:
> @brunodmartins thanks for the fix, can you please create another unittests for the case from the bug ticket so that we have this case covered forever
> 
> this unittest should lead to errors in the master and go through successfully in a pull request if you would transfer it for testing purposes

Sure! I pushed just to don't lose the job on my machine. I am coding the tests right now on some way it makes sense to exist, and not just to pass the coverage! I believe on the next 2 days I will open the PR for review.

brunodmartins commented:
@ReneWerner87 I have added the tests to validate the issue. I tested it before the fix and they were falling. Despite that, I added new tests due to coverage failure on the parts that were moved to the conditional block to be executed only in case the cache entry exist.

I need some help to validate why the tests are failing on Windows machines (PR actions) =(.

brunodmartins commented:
@gaby , I have checked your point that the `cache.go` should have some mutex to avoid Race conditions, but the calls to manipulate the cache are already secured by a mutex at the middleware level (a.k.a the caller). Could you do a double check? In case we really need to move the mutex more inside, I can try to work on it at another PR =)

brunodmartins commented:
Thanks again for the help and patience

ReneWerner87 commented:
> @ReneWerner87 I have added the tests to validate the issue. I tested it before the fix and they were falling. Despite that, I added new tests due to coverage failure on the parts that were moved to the conditional block to be executed only in case the cache entry exist.
> 
> I need some help to validate why the tests are failing on Windows machines (PR actions) =(.

![image](https://github.com/user-attachments/assets/03209156-359e-45b3-bb63-c4a39259a0bc)

you should not use the time as body, this varies greatly in the github machines used for the actions

rather use an incremental counter that counts up when the handler is reached

ReneWerner87 commented:
@brunodmartins just seen that the other test cases also use times as returns, but in the nano seconds range

can you do me a favor and rework them too, so that they use a variable that just counts up and gives a different result if there is no cache hit?

brunodmartins commented:
@ReneWerner87 thanks for the help! I was just that! I fixed the whole test file and it worked =)

sixcolors commented:
> @brunodmartins just seen that the other test cases also use times as returns, but in the nano seconds range ![image](https://private-user-images.githubusercontent.com/7063188/350438407-50a42eb0-d222-4ae0-af7f-20bc3eb4ec19.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE0ODA1NzQsIm5iZiI6MTcyMTQ4MDI3NCwicGF0aCI6Ii83MDYzMTg4LzM1MDQzODQwNy01MGE0MmViMC1kMjIyLTRhZTAtYWY3Zi0yMGJjM2ViNGVjMTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MjBUMTI1NzU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzgzY2Q4M2I1NmViZGFjZGEwY2MyMzk5NDllNWY4NzE1ZjdiYjlhYjUyY2EwNjRiMWFjMTBmNDdlNTE0NDY2YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Ra9HVbfDM9yh22bCRt9n9Tk6LrgDQn0piJUwb2HnGp0)
> 
> can you do me a favor and rework them too, so that they use a variable that just counts up and gives a different result if there is no cache hit?
> 
> all things with timing cause flakyness in our tests

"Timer resolution varies depending on the Go runtime, the operating system and the underlying hardware. On Unix, the resolution is approximately 1ms. On Windows, the default resolution is approximately 16ms, but a higher resolution may be requested using [golang.org/x/sys/windows.TimeBeginPeriod](https://pkg.go.dev/golang.org/x/sys/windows#TimeBeginPeriod)."
https://pkg.go.dev/time#hdr-Timer_Resolution

ReneWerner87 commented:
@gofiber/maintainers can you re-review

ReneWerner87 commented on middleware/cache/manager_test.go:
```diff
@@ -0,0 +1,23 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_manager_get(t *testing.T) {
```
@brunodmartins pls add t.Parallel() in parent and child tests

brunodmartins replied:
Done =)
